### PR TITLE
SASS-8883 add adjustments to the task list

### PR DESCRIPTION
**Pull request by kuldeepwadhwa9**

### CODE REVIEW

File: app/viewmodels/journeys/taskList/NationalInsuranceContributionsViewModel.scala [hunk 1]
```diff
@@ -26,22 +28,28 @@ import viewmodels.journeys.{SummaryListCYA, getJourneyStatus}
 
 object NationalInsuranceContributionsViewModel {
 
-  def buildSummaryList(nationalInsuranceStatuses: List[JourneyNameAndStatus])(implicit messages: Messages): SummaryList = {
+  def buildSummaryList(
+      nationalInsuranceStatuses: List[JourneyNameAndStatus])(implicit messages: Messages, taxYear: TaxYear, businessId: BusinessId): SummaryList = {
 
-    val nicRow = buildRow(NationalInsuranceContributions, nationalInsuranceStatuses, dependentJourneyIsFinishedForClickableLink = false)
+    val nicRow = buildRow(NationalInsuranceContributions, nationalInsuranceStatuses, dependentJourneyIsFinishedForClickableLink = false)(
```
Comment (Jake-Raffe) [inline app/viewmodels/journeys/taskList/NationalInsuranceContributionsViewModel.scala:34]: dependentJourneyIsFinishedForClickableLink will be dependent on all the previous statuses being 'Completed'

File: app/viewmodels/journeys/taskList/NationalInsuranceContributionsViewModel.scala [hunk 2]
```diff
@@ -26,22 +28,28 @@ import viewmodels.journeys.{SummaryListCYA, getJourneyStatus}
 
 object NationalInsuranceContributionsViewModel {
 
-  def buildSummaryList(nationalInsuranceStatuses: List[JourneyNameAndStatus])(implicit messages: Messages): SummaryList = {
+  def buildSummaryList(
+      nationalInsuranceStatuses: List[JourneyNameAndStatus])(implicit messages: Messages, taxYear: TaxYear, businessId: BusinessId): SummaryList = {
```
Comment (Jake-Raffe) [inline app/viewmodels/journeys/taskList/NationalInsuranceContributionsViewModel.scala:32]: taxYear doesn't need to be implicit, and we don't need businessId (you might need to remove it from anywhere it businessId is used for that page in the controller and the routes)

File: app/viewmodels/journeys/taskList/TradeJourneyStatusesViewModel.scala [hunk 3]
```diff
@@ -48,10 +48,14 @@ object TradeJourneyStatusesViewModel {
     val isAbroadAnswered = tradesJourneyStatuses.getStatusOrNotStarted(Abroad).isCompleted
     val incomeRow        = buildRow(Income, dependentJourneyIsFinishedForClickableLink = isAbroadAnswered)
 
+    val isCarAndAssetsBasedAllowanceAnswered = tradesJourneyStatuses.getStatusOrNotStarted(CapitalAllowancesTailoring).isCompleted
+    val adjustmentsRow = buildRow(Adjustments, dependentJourneyIsFinishedForClickableLink = isCarAndAssetsBasedAllowanceAnswered)
```
Comment (Jake-Raffe) [inline app/viewmodels/journeys/taskList/TradeJourneyStatusesViewModel.scala:52]: dependent on expenses AND capital allowances

File: app/viewmodels/journeys/taskList/TradeJourneyStatusesViewModel.scala [hunk 4]
```diff
@@ -64,9 +68,10 @@ object TradeJourneyStatusesViewModel {
     val status: JourneyStatus = getJourneyStatus(journey, dependentJourneyIsFinishedForClickableLink)(journeyStatuses.journeyStatuses)
     val keyString             = messages(s"journeys.$journey")
     val href = journey match {
-      case Abroad => getAbroadUrl(status, businessId, taxYear)
-      case Income => getIncomeUrl(status, businessId, taxYear)
-      case _      => "#"
+      case Abroad      => getAbroadUrl(status, businessId, taxYear)
+      case Income      => getIncomeUrl(status, businessId, taxYear)
+      case Adjustments => "#"
+      case _           => "#"
```
Comment (Jake-Raffe) [inline app/viewmodels/journeys/taskList/TradeJourneyStatusesViewModel.scala:74]: I'd make a 'getAdjustmentsUrl(status, businessId, taxYear)' method at the bottom like the others and put the "#" in the method

File: app/viewmodels/journeys/taskList/NationalInsuranceContributionsViewModel.scala (hunk 1 above)
Comment (kuldeepwadhwa9) [reply app/viewmodels/journeys/taskList/NationalInsuranceContributionsViewModel.scala:34]: Done :)

File: app/viewmodels/journeys/taskList/TradeJourneyStatusesViewModel.scala (hunk 3 above)
Comment (kuldeepwadhwa9) [reply app/viewmodels/journeys/taskList/TradeJourneyStatusesViewModel.scala:52]: Done :)

File: app/viewmodels/journeys/taskList/TradeJourneyStatusesViewModel.scala (hunk 4 above)
Comment (kuldeepwadhwa9) [reply app/viewmodels/journeys/taskList/TradeJourneyStatusesViewModel.scala:74]: Done :)

File: app/viewmodels/journeys/taskList/NationalInsuranceContributionsViewModel.scala (hunk 2 above)
Comment (kuldeepwadhwa9) [reply app/viewmodels/journeys/taskList/NationalInsuranceContributionsViewModel.scala:32]: Done :)

File: app/viewmodels/journeys/taskList/NationalInsuranceContributionsViewModel.scala [hunk 5]
```diff
@@ -26,22 +28,27 @@ import viewmodels.journeys.{SummaryListCYA, getJourneyStatus}
...
 
-  def buildSummaryList(nationalInsuranceStatuses: List[JourneyNameAndStatus])(implicit messages: Messages): SummaryList = {
+  def buildSummaryList(nationalInsuranceStatuses: List[JourneyNameAndStatus], taxYear: TaxYear)(implicit messages: Messages): SummaryList = {
 
-    val nicRow = buildRow(NationalInsuranceContributions, nationalInsuranceStatuses, dependentJourneyIsFinishedForClickableLink = false)
+    val isAdjustmentsAnswered = JourneyStatus.getJourneyStatus(Adjustments, nationalInsuranceStatuses).isCompleted
+    val nicRow = buildRow(
+      NationalInsuranceContributions,
+      nationalInsuranceStatuses,
+      dependentJourneyIsFinishedForClickableLink = isAdjustmentsAnswered)(messages, taxYear)
```
Comment (Jake-Raffe) [inline app/viewmodels/journeys/taskList/NationalInsuranceContributionsViewModel.scala:37]: taxyear can be passed in as a no implicit argument, which means messages also doesn't need to be passed in the 2nd brackets (will do so implicitly)

File: app/viewmodels/journeys/taskList/NationalInsuranceContributionsViewModel.scala [hunk 6]
```diff
@@ -26,22 +28,27 @@ import viewmodels.journeys.{SummaryListCYA, getJourneyStatus}
 
 object NationalInsuranceContributionsViewModel {
 
-  def buildSummaryList(nationalInsuranceStatuses: List[JourneyNameAndStatus])(implicit messages: Messages): SummaryList = {
+  def buildSummaryList(nationalInsuranceStatuses: List[JourneyNameAndStatus], taxYear: TaxYear)(implicit messages: Messages): SummaryList = {
 
-    val nicRow = buildRow(NationalInsuranceContributions, nationalInsuranceStatuses, dependentJourneyIsFinishedForClickableLink = false)
+    val isAdjustmentsAnswered = JourneyStatus.getJourneyStatus(Adjustments, nationalInsuranceStatuses).isCompleted
```
Comment (Jake-Raffe) [inline app/viewmodels/journeys/taskList/NationalInsuranceContributionsViewModel.scala:33]: Adjustments isn't a part of nationalInsuranceStatuses. 
There will only ever be one status for national insurance, so we can change this in the tasklist model in the FE and BE from `nationalInsuranceStatuses: List[JourneyNameAndStatus]` to just `nationalInsuranceStatus: JourneyNameAndStatus`.

We'll then need to pass in the completedTrades into this method from the TaskListController to check if Adjustments is completed for each business

File: app/viewmodels/journeys/taskList/NationalInsuranceContributionsViewModel.scala [hunk 7]
```diff
@@ -26,22 +28,27 @@ import viewmodels.journeys.{SummaryListCYA, getJourneyStatus}
...
-  private def buildRow(journey: Journey, nationalInsuranceStatuses: List[JourneyNameAndStatus], dependentJourneyIsFinishedForClickableLink: Boolean)(
-      implicit messages: Messages): SummaryListRow = {
+  private def buildRow(journey: Journey,
+                       nationalInsuranceStatuses: List[JourneyNameAndStatus],
+                       dependentJourneyIsFinishedForClickableLink: Boolean)(implicit messages: Messages, taxYear: TaxYear): SummaryListRow = {
     val status: JourneyStatus = getJourneyStatus(journey, dependentJourneyIsFinishedForClickableLink)(nationalInsuranceStatuses)
     val keyString             = messages(s"journeys.$journey")
     val href = journey match {
-      case NationalInsuranceContributions => "#"
+      case NationalInsuranceContributions => nics.routes.Class2NICsController.onPageLoad(taxYear, NormalMode).url
```
Comment (Jake-Raffe) [inline app/viewmodels/journeys/taskList/NationalInsuranceContributionsViewModel.scala:51]: Since we know which journey this method will make, we don't need to pass in the journey.
the keyString can just be `messages(s"journeys.$NationalInsuranceContributions")` and we don't need to match the journey for the href.

File: app/viewmodels/journeys/taskList/TradeJourneyStatusesViewModel.scala [hunk 8]
```diff
@@ -48,14 +48,26 @@ object TradeJourneyStatusesViewModel {
     val isAbroadAnswered = tradesJourneyStatuses.getStatusOrNotStarted(Abroad).isCompleted
     val incomeRow        = buildRow(Income, dependentJourneyIsFinishedForClickableLink = isAbroadAnswered)
 
+    val expensesRows: Seq[SummaryListRow] = buildExpensesCategories
+    val expensesAllCompleted: Boolean     = expensesRows.forall(checkIfRowIsCompleted)
+
+    val capitalAllowanceRows: Seq[SummaryListRow] = buildCapitalAllowances(tradesJourneyStatuses, taxYear)
+    val capitalAllowanceAllCompleted: Boolean     = capitalAllowanceRows.forall(checkIfRowIsCompleted)
+
+    val adjustmentsRow = buildRow(Adjustments, dependentJourneyIsFinishedForClickableLink = capitalAllowanceAllCompleted && expensesAllCompleted)
```
Comment (Jake-Raffe) [inline app/viewmodels/journeys/taskList/TradeJourneyStatusesViewModel.scala:57]: Needs to check income as well because income could be incomplete meaning expenses would be empty but still pass true

File: conf/messages.en [hunk 9]
```diff
@@ -93,6 +93,7 @@ journeys.capital-allowances-annual-investment-allowance = Annual Investment Allo
 journeys.capital-allowances-writing-down-allowance = ‘Writing down’ allowance
 journeys.capital-allowances-special-tax-sites = Freeport sites
 journeys.capital-allowances-structures-buildings = Structures and Buildings Allowance
+journeys.adjustments-work-out-profit-and-loss = Work out your adjusted profit or loss
```
Comment (Jake-Raffe) [inline conf/messages.en:96]: This is the text for Journey.ProfitOrLoss so should be journey.profit-or-loss

File: test/viewmodels/NationalInsuranceContributionsViewModelSpec.scala [hunk 10]
```diff
@@ -32,16 +34,18 @@ class NationalInsuranceContributionsViewModelSpec extends SpecBase with TableDri
 
   private implicit val messages: Messages = messagesStubbed
 
+  private val nicUrl = nics.routes.Class2NICsController.onPageLoad(taxYear, NormalMode).url
+
   private val testScenarios = Table(
     ("nationalInsuranceStatuses", "userAnswers", "expected"),
     // No statuses, no answers, defaults to cannot start yet until the saving is implemented
-    (Nil, Nil, List(expectedRow("#", NationalInsuranceContributions, CannotStartYet)))
+    (Nil, Nil, List(expectedRow(nicUrl, NationalInsuranceContributions, CannotStartYet)))
```
Comment (Jake-Raffe) [inline test/viewmodels/NationalInsuranceContributionsViewModelSpec.scala:42]: Replace "userAnswers" which isn't being used for "businessStatuses" (or a name similar) for the PR comment I added in NationalInsuranceContributionsViewModel.

Can then replace the comment above for comments and tests for:
- No businessStatuses but test result for each NIC status
- How the various statuses are influenced if adjustments in businessStatuses are different, and if they're different for different businesses

File: app/models/journeys/Journey.scala [hunk 11]
```diff
@@ -313,6 +313,11 @@ object Journey extends Enum[Journey] with utils.PlayJsonEnum[Journey] {
   sealed abstract class AdjustmentsBaseJourney(override val entryName: String) extends Journey(entryName) {
     override def toString: String = entryName
   }
+
+  case object Adjustments extends AdjustmentsBaseJourney("adjustments-work-out-profit-and-loss") {
+    override val pageKeys: List[PageName] = List.empty
```
[the remark this answers is not in the document]
Comment (Jake-Raffe) [reply app/models/journeys/Journey.scala:318]: Sorry thought this was the base journey but it isn't. We don't need the case object Adjustments because we have profit and loss below, so this object can just be deleted

File: app/viewmodels/journeys/taskList/NationalInsuranceContributionsViewModel.scala (hunk 5 above)
Comment (kuldeepwadhwa9) [reply app/viewmodels/journeys/taskList/NationalInsuranceContributionsViewModel.scala]: Done :)

File: app/viewmodels/journeys/taskList/NationalInsuranceContributionsViewModel.scala (hunk 6 above)
Comment (kuldeepwadhwa9) [reply app/viewmodels/journeys/taskList/NationalInsuranceContributionsViewModel.scala]: Done :)

File: app/viewmodels/journeys/taskList/NationalInsuranceContributionsViewModel.scala (hunk 7 above)
Comment (kuldeepwadhwa9) [reply app/viewmodels/journeys/taskList/NationalInsuranceContributionsViewModel.scala]: Done :)

File: app/models/journeys/Journey.scala (hunk 11 above)
[the remark this answers is not in the document]
Comment (kuldeepwadhwa9) [reply app/models/journeys/Journey.scala:318]: Done :)

File: conf/messages.en (hunk 9 above)
Comment (kuldeepwadhwa9) [reply conf/messages.en]: Done :)

File: test/viewmodels/NationalInsuranceContributionsViewModelSpec.scala (hunk 10 above)
Comment (kuldeepwadhwa9) [reply test/viewmodels/NationalInsuranceContributionsViewModelSpec.scala]: Done :)

File: app/viewmodels/journeys/taskList/TradeJourneyStatusesViewModel.scala (hunk 8 above)
Comment (kuldeepwadhwa9) [reply app/viewmodels/journeys/taskList/TradeJourneyStatusesViewModel.scala]: Done :)

File: test/viewmodels/NationalInsuranceContributionsViewModelSpec.scala [hunk 12]
```diff
@@ -36,16 +37,92 @@ class NationalInsuranceContributionsViewModelSpec extends SpecBase with TableDri
...
+
+  private val nicCompleteStatus: Option[JourneyNameAndStatus] = Some(JourneyNameAndStatus(NationalInsuranceContributions, JourneyStatus.Completed))
+
   private val testScenarios = Table(
-    ("nationalInsuranceStatuses", "userAnswers", "expected"),
-    // No statuses, no answers, defaults to cannot start yet until the saving is implemented
-    (Nil, Nil, List(expectedRow(nicUrl, NationalInsuranceContributions, CannotStartYet)))
+    ("nationalInsuranceStatus", "businessStatuses", "expected"),
+    // If Business status does not contain completed adjustment status; NIC status should be CanNotStartYet
+    (
```
Comment (Jake-Raffe) [inline test/viewmodels/NationalInsuranceContributionsViewModelSpec.scala:52]: if adjustments in businessStatuses are not ALL completed then the UI status will be CannotStartYet regardless of the saved NICStatus.

1. None / List(Model(List(Adj, InProgress))) -> CannotStartYet
2. NotStarted / List(Model(List(Adj, Completed)), Model(List(Adj, Completed)), Model(List(Adj, CSY))) -> CSY
3. InProgress / List(Model(List(Adj, Completed)), Model(List(Adj, InProgress))) -> CSY
4. Completed /  List(Model(List.empty)) -> CSY

File: test/viewmodels/NationalInsuranceContributionsViewModelSpec.scala [hunk 13]
```diff
@@ -36,16 +37,92 @@ class NationalInsuranceContributionsViewModelSpec extends SpecBase with TableDri
...
+          tradingName = Some(TradingName("TradingName2")),
+          journeyStatuses = List(JourneyNameAndStatus(Adjustments, JourneyStatus.NotStarted))),
+        aTadesJourneyStatusesModel.copy(
+          tradingName = Some(TradingName("TradingName3")),
+          journeyStatuses = List(JourneyNameAndStatus(Adjustments, JourneyStatus.InProgress)))
+      ),
+      List(expectedRow(nicUrl, NationalInsuranceContributions, CannotStartYet))),
+
+    // If Adjustments is completed, it should show NIC status as NotStarted
+    (
```
Comment (Jake-Raffe) [inline test/viewmodels/NationalInsuranceContributionsViewModelSpec.scala:87]: if adjustments in businessStatuses are ALL completed then the UI status will match the saved NICStatus or default to NotStarted.

1. None / List(Model(List(Adj, Completed))) -> NotStarted
2. NotStarted / List(Model(List(Adj, Completed)), Model(List(Adj, Completed))) -> NS
3. InProgress / List(Model(List(Adj, Completed))) -> IP
4. Completed / List(Model(List(Adj, Completed)), Model(List(Adj, Completed))) -> Completed

**Review by Jake-Raffe — APPROVED** [review]
